### PR TITLE
Warnings

### PR DESCRIPTION
**Pull request by mhauru**
The basic framework is there, a few warning types are successfully produced and displayed on the home page. More polish is needed for the front end, as are more warning types. Adding more warning types may also warrant changing how the backend and/or database are structured, no reason to be too respectful of the current design. The function app hasn't been tested (I've been running `python3 -m core.warnings` locally), let's see how it fares after CD deploys it.

See https://github.com/alan-turing-institute/CROP/issues/173 for discussion.

### CODE REVIEW

**Review by nbarlowATI — COMMENTED** [review]
Code looks good!     I think you mentioned that we may want to have separation between warnings for developers and warnings for farm operators, but this is a good thing to build from!  
When I tried running though, I got the following error when loading the homepage:
```
127.0.0.1 - - [12/Sep/2022 14:16:59] "GET /home/index HTTP/1.1" 500 -
Traceback (most recent call last):
  File "/Users/nbarlow/anaconda3/envs/cropenv/lib/python3.8/site-packages/pandas/core/indexes/base.py", line 3361, in get_loc
    return self._engine.get_loc(casted_key)
  File "pandas/_libs/index.pyx", line 76, in pandas._libs.index.IndexEngine.get_loc
  File "pandas/_libs/index.pyx", line 108, in pandas._libs.index.IndexEngine.get_loc
  File "pandas/_libs/hashtable_class_helper.pxi", line 5198, in pandas._libs.hashtable.PyObjectHashTable.get_item
  File "pandas/_libs/hashtable_class_helper.pxi", line 5206, in pandas._libs.hashtable.PyObjectHashTable.get_item
KeyError: 'template_values'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/nbarlow/anaconda3/envs/cropenv/lib/python3.8/site-packages/pandas/core/frame.py", line 3751, in _set_item_mgr
    loc = self._info_axis.get_loc(key)
  File "/Users/nbarlow/anaconda3/envs/cropenv/lib/python3.8/site-packages/pandas/core/indexes/base.py", line 3363, in get_loc
    raise KeyError(key) from err
KeyError: 'template_values'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/nbarlow/anaconda3/envs/cropenv/lib/python3.8/site-packages/flask/app.py", line 2095, in __call__
    return self.wsgi_app(environ, start_response)
  File "/Users/nbarlow/anaconda3/envs/cropenv/lib/python3.8/site-packages/flask/app.py", line 2080, in wsgi_app
    response = self.handle_exception(e)
  File "/Users/nbarlow/anaconda3/envs/cropenv/lib/python3.8/site-packages/flask_cors/extension.py", line 165, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
  File "/Users/nbarlow/anaconda3/envs/cropenv/lib/python3.8/site-packages/flask/app.py", line 2077, in wsgi_app
    response = self.full_dispatch_request()
  File "/Users/nbarlow/anaconda3/envs/cropenv/lib/python3.8/site-packages/flask/app.py", line 1525, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/Users/nbarlow/anaconda3/envs/cropenv/lib/python3.8/site-packages/flask_cors/extension.py", line 165, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
  File "/Users/nbarlow/anaconda3/envs/cropenv/lib/python3.8/site-packages/flask/app.py", line 1523, in full_dispatch_request
    rv = self.dispatch_request()
  File "/Users/nbarlow/anaconda3/envs/cropenv/lib/python3.8/site-packages/flask/app.py", line 1509, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
  File "/Users/nbarlow/anaconda3/envs/cropenv/lib/python3.8/site-packages/flask_login/utils.py", line 277, in decorated_view
    return current_app.ensure_sync(func)(*args, **kwargs)
  File "/Users/nbarlow/CROP/webapp/app/home/routes.py", line 505, in index
    warnings = get_warnings(dt_from_daily)
  File "/Users/nbarlow/CROP/webapp/app/home/routes.py", line 397, in get_warnings
    warnings["template_values"] = warnings.apply(
  File "/Users/nbarlow/anaconda3/envs/cropenv/lib/python3.8/site-packages/pandas/core/frame.py", line 3602, in __setitem__
    self._set_item_frame_value(key, value)
  File "/Users/nbarlow/anaconda3/envs/cropenv/lib/python3.8/site-packages/pandas/core/frame.py", line 3742, in _set_item_frame_value
    self._set_item_mgr(key, arraylike)
  File "/Users/nbarlow/anaconda3/envs/cropenv/lib/python3.8/site-packages/pandas/core/frame.py", line 3754, in _set_item_mgr
    self._mgr.insert(len(self._info_axis), key, value)
  File "/Users/nbarlow/anaconda3/envs/cropenv/lib/python3.8/site-packages/pandas/core/internals/managers.py", line 1162, in insert
    block = new_block(values=value, ndim=self.ndim, placement=slice(loc, loc + 1))
  File "/Users/nbarlow/anaconda3/envs/cropenv/lib/python3.8/site-packages/pandas/core/internals/blocks.py", line 1937, in new_block
    check_ndim(values, placement, ndim)
  File "/Users/nbarlow/anaconda3/envs/cropenv/lib/python3.8/site-packages/pandas/core/internals/blocks.py", line 1979, in check_ndim
    raise ValueError(
ValueError: Wrong number of items passed 9, placement implies 1
```
which looks like it could be related?   (though maybe i have some old version of some package....   Will investigate a bit more...)

File: ingress_functions/warnings_func/function.json [hunk 1]
```diff
@@ -0,0 +1,12 @@
+{
+  "scriptFile": "__init__.py",
+  "entryPoint": "create_warnings",
+  "bindings": [
+    {
+        "name": "mytimer",
+        "type": "timerTrigger",
+        "direction": "in",
+        "schedule": "0 */1 * * *"
```
[the remark this answers is not in the document]
Comment (mhauru) [reply ingress_functions/warnings_func/function.json]: Good idea, done.

**Comment by mhauru** [comment]
I fixed the crash Nick reported above, and improved the front end. There are now checkboxes for which types of warnings are to be displayed, and colour coding for priority.

Comments welcome on the front end. Otherwise the main thing on my mind is to work on the different types of warnings, adding new ones, figuring out what exactly we want warnings for.

**Comment by nbarlowATI** [comment]
This looks good to me now - are you planning on more changes before we merge @mhauru or shall we go ahead and merge?  (And then perhaps do a `dev` -> `main` merge and then ask people what they think would be good alerts?)

**Comment by mhauru** [comment]
There's no great harm in it if you'd like to merge, but there's still one feature that the more I think about it the more I want to add it, namely a `category` column in the `warning_types` table, that groups different types into broader categories like "Farm conditions", "Sensor issues", "Modelling issues". The checkboxes would then be based on those. Otherwise by the time we are done there will be >10 different checkboxes for different warning types. I think I should be able to do that on Monday, and maybe we could then merge ahead of the Tuesday meeting?

**Comment by mhauru** [comment]
I implemented the feature described above, warning types are now grouped into bigger categories and the checkboxes on the home page are based on those. Currently the categories are "Missing data" and "Farm conditions". I now have nothing in mind that I want to imminently implement other than adding more warning types. Good time to merge, @nbarlowATI?

**Pull request merged**